### ISSUE TITLE
Games that have no source set and the platform is set to something that is not PC does not get added to Playnite Web

### ISSUE THREAD

**Issue by TheVNConnoisseur** [andrew-codes/playnite-web#675]
As title suggests, the regex check for the platform entry that was added back in version 8.3.0 if my memory serves right, made all of my games that have no platform set as PC or PlayStation (or Xbox) is not added to Playnite Web.
The logs from the latest version do confirm that ALL of the entries are received correctly, and when I try to add a dummy entry with a platform set to PC/PlayStation, the game gets added immediately, so I can confirm that from my side the configuration has been done correctly.

Checking with the Mongo database, all of the metadata from my library gets added (minus the games), so it's not like it cannot parse correctly the platform names.

**Comment by R1C3R21** [comment]
Oh, @andrew-codes I have the same issue as this user comments here. So I'll just comment in here.

**Comment by andrew-codes** [comment]
I'll try to review some of these issues this upcoming week.

**Comment by andrew-codes** [comment]
I may need some collective help on how to best approach this problem. Let me provide some additional context in the reasoning and mechanics of how it works today. Then I'll try to outline some directions that could be taken. Sorry for the long post. I'm certainly seeking feedback and help on implementing a solution to this class of problem.

## Additional Context

A game in Playnite may be available on many platforms; all of which are included as a relation to it. For this reason, Playnite Web cannot simply look at the platforms to see what platform a game's release is associated with. Instead, it uses the source to infer the platform, i.e., a PlayStation source means PlayStation. Even this is challenging, as some games are available on multiple PlayStation platforms, e.g., a game on both PS4 and PS5. In accommodate this, a release is created for each available platform matching the source.

This is also why not all platforms are as fully supported as others. I have implemented this for PC, PlayStation, and most likely Xbox. I have not verified with other platforms, such as Nintendo, etc.

Some sources are problematic, specifically emulators, as there are two platforms: the original and the platform it is played (emulated on PC). Currently, Playnite Web favors using PC in this scenario. The reason for this is to enable remote control of starting/stopping games. My personal use case for Playnite Web is for it to be a remote control to playing games in my game room. However, I recognize that this may not fit everyone as some may want the platform to say the original platform.

## Problem

Without a source, the only thing available is a list of platforms that the game could be associated with; again, not the one(s) that it is associated with. In this case, it falls back to use the PC platform. This is because there is not a way to infer which of the platforms is used to play the game. My suspicion in this case is that it does not publish the game because there is not an identified platform. I'm guessing that it cannot find the PC platform, which is the fall back, in the list of available platforms.

## Possible Solutions

> Note that having more information on the games that do not synchronize will be helpful. That will provide more insight into the type of game and what the intended platform is that it is normally played on.

### Option 1

This is the most straight forward way. It requires no new feature changes to Playnite Web. Update affected games in Playnite by adding the PC platform as one of the game's platforms. The relation is to a PC platform entity in Playnite. Note that multiple platform entities can share the same name, but this does not make them the same; each is unique with its own ID.

### Option 2

Allow users to specify the default platform when none can be inferred; by having no source or otherwise. This could be an option in the settings of Playnite Web's plugin. Note, this does not truly solve the problem. It is possible to have multiple games with no source that should have different platforms. This would not accommodate for this scenario.

### Option 3

Do not rely on sources or platforms from Playnite at all. Instead of inferring, consider creating a new relation to Playnite Games. This new related entity, ReleasePlatform (tbd name), allows the user to set one or more platforms that the game is played on. This would allow the user to set this on a game-by-game basis.

This requires a good bit of work. For one, we need to store this new related entity somewhere. Likely, this would be in the form of a sqlite database file stored in the extension's data directory. Introducing this also means initializing an empty database, as well as performing CRUD operations.

The bulk of the work is not here; rather it is in providing a UI to update this new relation in games. I do not **think** this field can be included in the normal "more details" UI for a game. Please correct me if I'm wrong on this though as this would be ideal. Instead, I think we would need to either have an additional More option to set this. Another option is to have a different UI, like some other plugins, which allows setting this value.

An additional drawback to this: this means that every game must be individually set with this new relation; otherwise, it will be excluded from syncing. To ease this burden, there would additionally need to be a way to set this value across multiple games. This could be the same UI mentioned in the preceding paragraph.

## Evaluating Trade-offs

I wholeheartedly think option 3 is the best. It makes the most sense and solves this problem and a whole class of other problems by inferring the platform. However, the amount of effort required feels steep and requires a Windows development environment; of which I have an extremely limited one currently and it is certainly not ideal enough to full-on develop XAML UI's. This excludes the harder barriers to surmount: my time and my skill set. I'm a C# engineer, but I use it to develop services as opposed to UI's. I'm not convinced this is a skill set that I want to hone compared to others in my free time from a career perspective. This makes the impact on my time even more, well, impactful.

### Collaborating with Others

Despite this, I am willing to try to guide others that may want to take this problem on and bring it to fruition. Although I don't want to spend my time on the problem, I'd be less stringent with my time when it comes to helping/collaborating with others. I do think that someone else taking this on is a big effort and would require a good bit of time from me to aid, in terms of reviews, direction taken, etc. Again, I may be more willing to do that.

If no one else is able to help take this on, then I'd recommend trying Option 1 and see how far that gets you.

**Comment by R1C3R21** [comment]
Hi @andrew-codes, thanks your detailed explanation. I was also looking at first at the third option when I was thinking of possible solutions to this problem, since I was kinda skeptical of the somewhat dirty way of assuming a game entity's platform.

For now I would wait a bit, since developing something with a new UI (I'm also a C# developer myself) when the author of Playnite is still deciding on what framework Josef Nemec is going to use for the next BIG update for Playnite (and that includes reworking the entire emulator handling process), which I recommend reading on his posts on his Patreon page (although some of them are payblocked, ~but those can be accessed freely on some places~).
From what I have been seeing, unless he has changed his discourse again, he is going to base his next version on WPF, so that's somewhat reassuring.

In conclusion, I'd say that I'd take option number 4, which is to wait until Josef releases publicly the first version of Playnite 12, and then see if applying option number 3 would be possible, but that's something that we will not be able to know until the end of next year.

**Comment by andrew-codes** [comment]
@R1C3R21 thanks for the update! That is an excellent idea. I afraid I'm aware, really at all, of Josef Nemec's ideas, plans or the future direction of Playnite. Given this information, it certainly is worth waiting to see what the next major version brings. Hopefully, option 1 works well enough to tide people over until a future time.

For others that are seeking an answer, please refer to option 1 from this [comment](https://github.com/andrew-codes/playnite-web/issues/675#issuecomment-2565875360).

I will close this issue for now. Going with option 3 is a new feature, so it makes sense to track it in a feature request issue instead of here.
